peekingduck/runner.py

```diff
@@ -29,54 +29,57 @@
-
-        config_updates_cli (:obj:`str`): config changes passed as part of the \
-        cli command sed to modify the node configurations direct from cli.
-
-        CUSTOM_NODE_PARENT_FOLDER (:obj:`str`): path to folder which contains \
-        custom nodes that users have created to be used with PeekingDuck. \
-        For more information on using custom nodes, please refer to \
-        `getting started <getting_started/03_custom_nodes.html>`_.
-
-        nodes (:obj:`list` of :obj:`Node`): if not using declarations via yaml, \
-        initialize by giving the node stack directly as a list.
-
+        run_config_path (:obj:`pathlib.Path` | :obj:`None`): If a path to
+            *run_config.yml* is provided, uses
+            :py:class:`DeclarativeLoader <peekingduck.declarative_loader.DeclarativeLoader>`
+            to load the YAML file according to PeekingDuck's specified schema
+            to obtain the declared nodes that would be sequentially initialized
+            and used to create the pipeline for running inference.
+        config_updates_cli (:obj:`str` | :obj:`None`): Configuration changes
+            passed as part of the CLI command used to modify the node
+            configurations directly from CLI.
+        custom_nodes_parent_subdir (:obj:`str` | :obj:`None`): Relative path to
+            a folder which contains custom nodes that users have created to be
+            used with PeekingDuck. For more information on using custom nodes,
+            please refer to
+            `Getting Started <getting_started/03_custom_nodes.html>`_.
+        nodes (:obj:`List[AbstractNode]` | :obj:`None`): If a list of nodes is
+            provided, initialize by the node stack directly.
     """
 
     def __init__(
         self,
-        RUN_PATH: Path = None,
+        run_config_path: Path = None,
         config_updates_cli: str = None,
-        CUSTOM_NODE_PARENT_FOLDER: str = None,
+        custom_nodes_parent_subdir: str = None,
         nodes: List[AbstractNode] = None,
     ):
         self.logger = logging.getLogger(__name__)
-
-        if not nodes and RUN_PATH:
-            # create Graph to run
-            self.node_loader = DeclarativeLoader(
-                RUN_PATH, config_updates_cli, CUSTOM_NODE_PARENT_FOLDER  # type: ignore
-            )
-            self.pipeline = self.node_loader.get_pipeline()
-        # If Runner given nodes, instantiated_nodes is created differently
-        else:
-            try:
-                self.pipeline = Pipeline(nodes)  # type: ignore
-            except ValueError as error:
-                self.logger.error(str(error))
-                sys.exit(1)
+        try:
+            if nodes:
+                # instantiated_nodes is created differently when given nodes
+                self.pipeline = Pipeline(nodes)
+            elif run_config_path and config_updates_cli and custom_nodes_parent_subdir:
```

Comment:
Is this logic still correct? In some cases `config_updates_cli` or `custom_nodes_parent_subdir` could be `None`?
To try with: 1) python PeekingDuck; 2) Python API; 3) changing node configs via CLI

Reply:
The following is a comparison of the truth tables between the old and proposed runner logic. Using old argument names for ease of comparison. The runner logic essentially only cares about `nodes` and `RUN_PATH` determine the branching, the rest depends on handles (and unhandled) exceptions. 

With the exception of `nodes=False`, `RUN_PATH=True`, and invalid `CUSTOM_NODE_PARENT_FOLDER` in **Python API** mode, the outcomes are equivalent. The refactor does not have the unhandled `TypeError` which the current implementation has when `nodes=False` and `RUN_PATH=False`.

**Current implementation**
| nodes | RUN_PATH | self.pipeline |
|-------|----------|---------------|
| True | True | Pipeline |
| True | False | Pipeline |
| False | True (valid config_updates_cli and CUSTOM_NODE_PARENT_FOLDER) | DeclarativeLoader |
| False | True (invalid config_updates_cli) | ValueError (from ast) |
| False | True (invalid CUSTOM_NODE_PARENT_FOLDER) | DeclarativeLoader but custom nodes won't work |
| False | False | Pipeline (unhandled TypeError)  |

**PR**
| nodes | RUN_PATH | self.pipeline     |
|-------|----------|-------------------|
| True | True | Pipeline |
| True | False | Pipeline |
| False | True (valid config_updates_cli and CUSTOM_NODE_PARENT_FOLDER) | DeclarativeLoader |
| False | True (invalid config_updates_cli or CUSTOM_NODE_PARENT_FOLDER) | ValueError |
| False | False | ValueError |

I can make the changes to make the PR refactor mimic the original **Python API mode** behavior: when `nodes=False`, `RUN_PATH=True`, and invalid `CUSTOM_NODE_PARENT_FOLDER`, the code runs but only default nodes and no custom nodes. However, I am not sure if users who use the Python API mode will stick to only default nodes or keep all nodes in the default directory/location.

Reply:
Thanks for the detailed explanation, I think it's fine to keep your current implementation in this PR. The scenario of `nodes=False`, `RUN_PATH=True`, and `invalid CUSTOM_NODE_PARENT_FOLDER` in *Python API* mode is quite unlikely as in our documentation for Python API, users are supposed to feed nodes into the Runner.